enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,297 @@
+  cluster.
+* This change does not introduce a new way for a client to discover
+  that a cluster is being upgraded.
+* This change does not introduce a new way for a client to discover
+  the cluster topology (single-node, HA, external control-plane, etc.).
+
+## Proposal
+
+Because we want arbitrary clients to be able to read and understand
+the data, we will use a standard Kubernetes ConfigMap.
+
+Because we want all OpenShift variants to publish the information in
+the same way and that the data should not be writable by API callers,
+we will specify the name and namespace for the ConfigMap.
+
+Because MicroShift exposes a limited number of namespaces by default,
+we need to choose from one of those. The most suitable available
+namespace is `openshift`, so we will create the ConfigMap with name
+`version` in the `openshift` namespace.
+
+This choice will mean updating the implementation in MicroShift to
+change the namespace from `kube-public`, but since the ConfigMap will
+be OpenShift-specific we want to avoid collisions that might be
+introduced by other distributions in the `kube-public` namespace.
+
+Because we want all OpenShift variants to report the same type of
+information, we will keep the contents of the ConfigMap brief and
+specify the data key names in this enhancement:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: version
+  namespace: openshift
+data:
+  product: ""
+  major: ""
+  minor: ""
+  version: ""
+```
+
+- `product` should hold a representation of the OpenShift variant
+  name. For example, `OCP` for OpenShift in standalone or external
+  control-plane configurations of all topologies and `MicroShift` for
+  MicroShift instances.
+- `major` should hold the major version number of the product as a
+  string. For example, for `4.12.0` the value would be `"4"`.
+- `minor` should hold the minor version number of the product as a
+  string. For example, for `4.12.0` the value would be `"12"`.
+- `version` should hold the complete version of the product as a
+  string. For example, for `4.12.0` the value would be `"4.12.0"`. The
+  version string may also include other qualifiers in nightly,
+  sprintly, CI, or hot-fix builds. For example,
+  `4.11.0-0.nightly-2022-06-23-153912`.
+
+The `version` string is meant for display in user interface
+situations.  The `major` and `minor` values are included separately so
+that API users do not have to parse the `version` string before
+applying rules based on the values.
+
+### Workflow Description
+
+#### Determining the Version of a Cluster via the API
+
+API clients can read the `openshift/version` ConfigMap contents in the
+normal way.
+
+#### Determining the Version of a Cluster via the CLI
+
+The `oc login` command reports the `message` value of the
+`openshift/motd` ConfigMap when a user successfully authenticates. The
+command should be extended to include the `version` string of the
+`openshift/version` ConfigMap, when it is present.
```

Comment:
This reads like only `oc login` would emit the version. Is that your intention or would `oc version` similarly be extended with this data? It includes much of this already.

```
$ oc version -o yaml
clientVersion:
  buildDate: "2022-05-07T07:15:04Z"
  compiler: gc
  gitCommit: 0bf70d5c6e686521c3001c6b79167079b46498fb
  gitTreeState: clean
  gitVersion: 4.11.0-202205070617.p0.g0bf70d5.assembly.stream-0bf70d5
  goVersion: go1.17.5
  major: ""
  minor: ""
  platform: linux/amd64
kustomizeVersion: v4.5.4
openshiftVersion: 4.10.18
releaseClientVersion: 4.11.0-0.nightly-arm64-2022-05-09-151004
serverVersion:
  buildDate: "2022-05-10T16:30:48Z"
  compiler: gc
  gitCommit: 3c28e7a79b58e78b4c1dc1ab7e5f6c6c2d3aedd3
  gitTreeState: clean
  gitVersion: v1.23.5+3afdacb
  goVersion: go1.17.5
  major: "1"
  minor: "23"
  platform: linux/amd64
```

Reply:
Oops, yes, I left that out. If the `oc` user can't see the ClusterVersion resource, I think it doesn't include the `openshiftVersion` value today, right?

Reply:
I've added `oc version`.

Reply:
Yeah, it's two calls. One hits an unauthenticated endpoint and the other fetches the clusterversion resource and the two datasources are merged in presentation.

```
GET https://api.sdodson.openshift.com:6443/version?timeout=32s
Response Body: {
  "major": "1",
  "minor": "23",
  "gitVersion": "v1.23.5+3afdacb",
  "gitCommit": "3c28e7a79b58e78b4c1dc1ab7e5f6c6c2d3aedd3",
  "gitTreeState": "clean",
  "buildDate": "2022-05-10T16:30:48Z",
  "goVersion": "go1.17.5",
  "compiler": "gc",
  "platform": "linux/amd64"
}
GET https://api.sdodson.openshift.com:6443/apis/config.openshift.io/v1/clusterversions/version
Response Body: {"apiVersion":"config.openshift.io/v1","kind":"ClusterVersion",...
```